### PR TITLE
Bugfig/invalid read adaptor filter

### LINKED ISSUE #229: Memory error in `rascal::AdaptorFilter<rascal::StructureManagerLammps, 2ul>::add_cluster<2ul>`
Here is the valgrind report:

```
Start  8: test_adaptor_filter
8: Test command: /usr/bin/valgrind "--error-exitcode=100" "--leak-check=full" "--show-leak-kinds=all" "--errors-for-leak-kinds=all" "--track-origins=yes" "--dsymutil=yes" "--suppressions=/local/scratch/mveit/codes/librascal/tests/valgrind.suppressions" "/local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter" "--" "--report_level=detailed" "--build_info=TRUE"
8: Test timeout computed to be: 10000000
8: ==30877== Memcheck, a memory error detector
8: ==30877== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
8: ==30877== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
8: ==30877== Command: /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter -- --report_level=detailed --build_info=TRUE
8: ==30877== 
8: ==30877== Invalid read of size 8
8: ==30877==    at 0x143A17: void rascal::AdaptorFilter<rascal::StructureManagerLammps, 2ul>::add_cluster<2ul>(rascal::StructureManager<rascal::StructureManagerLammps>::ClusterRef<2ul> const&) (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x143E4A: rascal::test_adaptor_filter::filter_2_test<rascal::FilterFixture<rascal::StructureManagerLammps, 2ul> >::test_method() (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x145647: void rascal::test_adaptor_filter::filter_2_test_invoker::run<rascal::FilterFixture<rascal::StructureManagerLammps, 2ul> >(boost::type<rascal::FilterFixture<rascal::StructureManagerLammps, 2ul> >*) (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x48BF2AD: ??? (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48BCB2C: boost::execution_monitor::catch_signals(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48BCBB6: boost::execution_monitor::execute(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48BCC8A: boost::execution_monitor::vexecute(boost::function<void ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48EA655: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48CD2E3: ??? (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48CD5FB: ??? (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48CD5FB: ??? (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48C47BE: boost::unit_test::framework::run(unsigned long, bool) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==  Address 0x4f8b9c0 is 0 bytes after a block of size 32 alloc'd
8: ==30877==    at 0x4838DEF: operator new(unsigned long) (vg_replace_malloc.c:344)
8: ==30877==    by 0x141D41: void std::vector<unsigned long, std::allocator<unsigned long> >::_M_realloc_insert<unsigned long const&>(__gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, unsigned long const&) (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x14397F: void rascal::AdaptorFilter<rascal::StructureManagerLammps, 2ul>::add_cluster<1ul>(rascal::StructureManager<rascal::StructureManagerLammps>::ClusterRef<1ul> const&) (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x143AE7: void rascal::AdaptorFilter<rascal::StructureManagerLammps, 2ul>::add_cluster<2ul>(rascal::StructureManager<rascal::StructureManagerLammps>::ClusterRef<2ul> const&) (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x143E4A: rascal::test_adaptor_filter::filter_2_test<rascal::FilterFixture<rascal::StructureManagerLammps, 2ul> >::test_method() (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x145647: void rascal::test_adaptor_filter::filter_2_test_invoker::run<rascal::FilterFixture<rascal::StructureManagerLammps, 2ul> >(boost::type<rascal::FilterFixture<rascal::StructureManagerLammps, 2ul> >*) (in /local/scratch/mveit/codes/librascal/build/tests/test_adaptor_filter)
8: ==30877==    by 0x48BF2AD: ??? (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48BCB2C: boost::execution_monitor::catch_signals(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48BCBB6: boost::execution_monitor::execute(boost::function<int ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48BCC8A: boost::execution_monitor::vexecute(boost::function<void ()> const&) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48EA655: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
8: ==30877==    by 0x48CD2E3: ??? (in /usr/lib/libboost_unit_test_framework.so.1.71.0)
```

To reproduce, run the tests with the above valgrind command or checkout #207 locally.

### PR DESCRIPTION
**Pull request by mastricker**
(Describe your pull request in 1-2 sentences)

Fix #229

Error was in the access for the `.back()` operator in `PropertyTyped`. The index calculation was not accounting for the _start at zero_.

I also removed the suppressions and left a comment in the test where the `random_device` is chosen to be able to quickly switch to predictable behaviour in the future.

### CODE REVIEW

File: src/rascal/structure_managers/property_typed.hh [hunk 1]
```diff
@@ -319,7 +319,11 @@ namespace rascal {
      * Accessor for last pushed entry for dynamically sized properties
      */
     reference back() {
-      auto && index{this->values.size() - this->get_nb_comp()};
+      // -1 is correction for 0-start indexing
+      auto && index{this->values.size() / this->get_nb_comp() - 1};
+      if (index < 0) {
```
Comment (Luthaf) [inline src/rascal/structure_managers/property_typed.hh]: if `this->values.size()` is 0, index will be -1/SIZE_MAX. So this check will not catch this case. Is it checked elsewhere?

File: src/rascal/structure_managers/property_typed.hh [hunk 2]
```diff
@@ -319,7 +319,11 @@ namespace rascal {
      * Accessor for last pushed entry for dynamically sized properties
      */
     reference back() {
-      auto && index{this->values.size() - this->get_nb_comp()};
+      // -1 is correction for 0-start indexing
+      auto && index{this->values.size() / this->get_nb_comp() - 1};
```
Comment (Luthaf) [inline src/rascal/structure_managers/property_typed.hh]: Previous code was `this->values.size() - this->get_nb_comp()`, now it is `this->values.size() / this->get_nb_comp() - 1`. Was the previous version completely bugged? The new one is very different.

File: src/rascal/structure_managers/property_typed.hh [hunk 3]
```diff
@@ -319,7 +319,11 @@ namespace rascal {
...
+      if (index < 0) {
+        throw std::runtime_error("Property is empty, .back() is undefined.");
```
Comment (Luthaf) [inline src/rascal/structure_managers/property_typed.hh]: Could you add a test that this error is thrown if the property is empty?

File: src/rascal/structure_managers/property_typed.hh (hunk 1 above)
Comment (mastricker) [reply src/rascal/structure_managers/property_typed.hh]: I'll switch the runtime check around and just test for an empty property.

File: src/rascal/structure_managers/property_typed.hh (hunk 2 above)
Comment (mastricker) [reply src/rascal/structure_managers/property_typed.hh]: My rationale was the following.
The previous one tried to get the index from the back, which did not correct for the zero offset. 
What I now check is how many entries are in the vector and then go back one entry. The access of the values is then `index * nb_comp`.

File: src/rascal/structure_managers/property_typed.hh (hunk 3 above)
Comment (mastricker) [reply src/rascal/structure_managers/property_typed.hh]: You mean forcibly test for the throwing of the runtime error?

File: src/rascal/structure_managers/property_typed.hh (hunk 2 above)
Comment (Luthaf) [reply src/rascal/structure_managers/property_typed.hh]: Oh, OK. The indexing code below did not changed and was already using `index * nb_comp`, so I guess it was very buggy.

File: src/rascal/structure_managers/property_typed.hh (hunk 3 above)
Comment (Luthaf) [reply src/rascal/structure_managers/property_typed.hh]: yes, using `BOOST_CHECK_THROWS` https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/boost_test/utf_reference/testing_tool_ref/assertion_boost_level_throw.html

Comment (mastricker) [reply src/rascal/structure_managers/property_typed.hh]: Sure

File: src/rascal/structure_managers/property_typed.hh (hunk 2 above)
Comment (Luthaf) [reply src/rascal/structure_managers/property_typed.hh]: This is slightly worrying me, since the behaviour changed a lot here but tests still pass. I guess this part was not tested.

**Comment by Luthaf** [comment]
> Error was in the access for the .back() operator in PropertyTyped. The index calculation was not accounting for the start at zero.

There are two hard things in programming: naming, cache invalidation and off-by-one errors.

**Comment by mastricker** [comment]
The .back() function is actually only used in the adaptor filter. So there should not be any issue anywhere else as far as I looked. (I had the same concern)

On Dec 6, 2019 11:46, Guillaume Fraux <notifications@github.com> wrote:

@Luthaf commented on this pull request.

________________________________

In src/rascal/structure_managers/property_typed.hh<https://github.com/cosmo-epfl/librascal/pull/240#discussion_r354770555>:

> @@ -319,7 +319,11 @@ namespace rascal {
      * Accessor for last pushed entry for dynamically sized properties
      */
     reference back() {
-      auto && index{this->values.size() - this->get_nb_comp()};
+      // -1 is correction for 0-start indexing
+      auto && index{this->values.size() / this->get_nb_comp() - 1};


This is slightly worrying me, since the behaviour changed a lot here but tests still pass. I guess this part was not tested.

-

**Comment by mastricker** [comment]
I changed the runtime error to check for the size of the vector and added a corresponding runtime error test for `PropertyTyped`.

File: src/rascal/structure_managers/property_typed.hh [hunk 4]
```diff
@@ -319,7 +319,11 @@ namespace rascal {
      * Accessor for last pushed entry for dynamically sized properties
      */
     reference back() {
-      auto && index{this->values.size() - this->get_nb_comp()};
+      // -1 is correction for 0-start indexing
```
Comment (Luthaf) [inline src/rascal/structure_managers/property_typed.hh]: The comment should move next to the `index = ...` line

Comment (mastricker) [reply src/rascal/structure_managers/property_typed.hh]: Sure, I'll wait for the CI to pass again.
